pkgs/build-support/make-desktopitem/default.nix

```
@@ -1,67 +1,96 @@
-, extraDesktopEntries ? { } # Extra key-value pairs to add to the [Desktop Entry] section. This may override other values
-, extraEntries ? "" # Extra configuration. Will be appended to the end of the file and may thus contain extra sections
-, fileValidation ? true # whether to validate resulting desktop file.
+# not supported until version 1.5, which is not supported by our desktop-file-utils
+# , singleMainWindow ? null
+, extraConfig ? {}
 }:
 let
-  # like builtins.toString, but null -> null instead of null -> ""
-  nullableToString = value:
+  boolOrNullToString = value:
     if value == null then null
     else if builtins.isBool value then lib.boolToString value
-    else builtins.toString value;
+    else throw "Value must be a boolean or null!";
+
+  renderList = value:
+    if !builtins.isList value then throw "Value must be a list!"
+    else if builtins.any (item: lib.hasInfix ";" item) value then throw "Values in list must not contain semicolons!"
+    else if value == [] then null
+    else builtins.concatStringsSep ";" value;
 
   # The [Desktop entry] section of the desktop file, as attribute set.
   mainSection = {
-    "Type" = toString type;
-    "Exec" = nullableToString exec;
-    "Icon" = nullableToString icon;
-    "Comment" = nullableToString comment;
-    "Terminal" = nullableToString terminal;
-    "Name" = toString desktopName;
-    "GenericName" = nullableToString genericName;
-    "MimeType" = nullableToString mimeType;
-    "Categories" = nullableToString categories;
-    "StartupNotify" = nullableToString startupNotify;
-    "NoDisplay" = nullableToString noDisplay;
-    "PrefersNonDefaultGPU" = nullableToString prefersNonDefaultGPU;
-  } // extraDesktopEntries;
+    "Type" = type;
+    "Version" = "1.4";
+    "Name" = desktopName;
+    "GenericName" = genericName;
+    "NoDisplay" = boolOrNullToString noDisplay;
+    "Comment" = comment;
+    "Icon" = icon;
+    "OnlyShowIn" = renderList onlyShowIn;
+    "NotShowIn" = renderList notShowIn;
+    "DBusActivatable" = boolOrNullToString dbusActivatable;
+    "TryExec" = tryExec;
+    "Exec" = exec;
+    "Path" = path;
+    "Terminal" = boolOrNullToString terminal;
+    "Actions" = renderList (builtins.attrNames actions);
+    "MimeType" = renderList mimeTypes;
+    "Categories" = renderList categories;
+    "Implements" = renderList implements;
+    "Keywords" = renderList keywords;
+    "StartupNotify" = boolOrNullToString startupNotify;
+    "StartupWMClass" = startupWMClass;
+    "URL" = url;
+    "PrefersNonDefaultGPU" = boolOrNullToString prefersNonDefaultGPU;
+    # "SingleMainWindow" = boolOrNullToString singleMainWindow;
+  } // extraConfig;
+
+  renderLine = name: value: if value != null then "${name}=${value}" else null;
+
+  renderSection = sectionName: attrs:
+    let
+      lines = builtins.filter (v: v != null) (lib.mapAttrsToList renderLine attrs);
+    in
+      builtins.concatStringsSep "\n" ([ "[${sectionName}]" ] ++ lines);
```

IMO use of `pipe` here would be easier to understand:
```suggestion
    lib.pipe attrs [
      (lib.mapAttrsToList renderLine)
      (builtins.filter (v: !isNull v))
      (builtins.concatStringsSep "\n")
      (section: ''
        [${sectionName}]
        ${section}
      '')
    ];
```

---

Thanks, I do like this more. Applied.